R/CohortSample.R

```diff
@@ -0,0 +1,285 @@
+#' @param outputDatabaseSchema  optional schema to output cohorts to (if different from cohortDatabaseSchema)
+#' @export
+#' @returns                     sampledCohortDefinitionSet - a data.frame like object that contains the resulting identifiers and modified names of cohorts
+#' @inheritParams               generateCohortSet
+sampleCohortDefinitionSet <- function(cohortDefinitionSet,
+                                      cohortIds = cohortDefinitionSet$cohortId,
+                                      connectionDetails = NULL,
+                                      connection = NULL,
+                                      tempEmulationSchema = getOption("sqlRenderTempEmulationSchema"),
+                                      cohortDatabaseSchema,
+                                      outputDatabaseSchema = cohortDatabaseSchema,
+                                      cohortTableNames = getCohortTableNames(),
+                                      n = NULL,
+                                      sampleFraction = NULL,
+                                      seed = 64374,
+                                      seedArgs = NULL,
+                                      identifierExpression = "cohortId * 1000 + seed",
+                                      incremental = FALSE,
+                                      incrementalFolder = NULL) {
+
+  checkmate::assertIntegerish(n, len = 1, null.ok = TRUE)
+  checkmate::assertNumeric(sampleFraction, len = 1, null.ok = TRUE, lower = 0, upper = 1.0)
+  checkmate::assertIntegerish(seed, min.len = 1)
+  checkmate::assertDataFrame(cohortDefinitionSet, min.rows = 1, col.names = "named")
+  checkmate::assertNames(colnames(cohortDefinitionSet),
+                         must.include = c(
+                           "cohortId",
+                           "cohortName",
+                           "sql"
+                         )
+  )
+
+  if (is.null(n) && is.null(sampleFraction))
+    stop("Must specificy n or fraction size")
+
+  if (is.null(connection) && is.null(connectionDetails)) {
+    stop("You must provide either a database connection or the connection details.")
+  }
+
+  if (incremental) {
+    if (is.null(incrementalFolder)) {
+      stop("Must specify incrementalFolder when incremental = TRUE")
+    }
+    if (!file.exists(incrementalFolder)) {
+      dir.create(incrementalFolder, recursive = TRUE)
+    }
+
+    recordKeepingFile <- file.path(incrementalFolder, "GeneratedCohortSamples.csv")
+  }
+  # check uniqueness of output ids
+  .checkUniqueOutputIds(cohortDefinitionSet$cohortIds, seed, identifierExpression, cohortTableNames)
+
+  start <- Sys.time()
+  if (is.null(connection)) {
+    connection <- DatabaseConnector::connect(connectionDetails)
+    on.exit(DatabaseConnector::disconnect(connection))
+  }
+
+  .checkCohortTables(connection, cohortDatabaseSchema, cohortTableNames)
+  sampledCohorts <-
+    purrr::map2(seed, cohortIds, function(seed, targetCohortId) {
```

Comment:
Could we instead do this operation in a loop instead of introducing a dependency on `purrr` to essentially bind rows of a data table?